activerecord/lib/active_record/autosave_association.rb

```diff
@@ -456,9 +456,17 @@ def save_has_one_association(reflection)
 
       # If the record is new or it has changed, returns true.
       def record_changed?(reflection, record, key)
-        record.new_record? ||
-          (record.has_attribute?(reflection.foreign_key) && record[reflection.foreign_key] != key) ||
-          record.will_save_change_to_attribute?(reflection.foreign_key)
+        return true if record.new_record?
+        return true if _association_foreign_key_changed?(reflection, record, key)
+        return true if record.will_save_change_to_attribute?(reflection.foreign_key)
+
+        false
+      end
+
+      def _association_foreign_key_changed?(reflection, record, key)
+        return false if reflection.through_reflection
```

Comment:
Use `through_reflection?`.

```suggestion
        !reflection.through_reflection? &&
          record.has_attribute?(reflection.foreign_key) && record[reflection.foreign_key] != key
```

Reply:
I dared to leave this as is. Comparing with the other case, where are 3 guards, here is only one. 
And mentally things are different enough to not join them into a single complex condition phrase. I consider this as a mentally easier, however less concise